src/renderer/store/handlers/directMessages.ts

```diff
@@ -0,0 +1,230 @@
+import { produce, immerable } from 'immer'
+import crypto from 'crypto'
+import { createAction, handleActions } from 'redux-actions'
+
+import directMessagesSelectors from '../selectors/directMessages'
+import identitySelectors from '../selectors/identity'
+import channelSelectors from '../selectors/channel'
+
+import { actionTypes } from '../../../shared/static'
+import { ActionsType, PayloadType } from './types'
+import { directMessagesActions } from '../../sagas/directMessages/directMessages.reducer'
+
+interface IUser {
+  nickname: string
+  publicKey: string
+  halfKey: string
+}
+
+interface IConversation {
+  sharedSecret: string
+  contactPublicKey: string
+  conversationId: string
+}
+
+export class DirectMessages {
+  users: { [key: string]: IUser }
+  conversations: { [key: string]: IConversation }
+  conversationsList: { [key: string]: string }
+  publicKey: string
+  privateKey: string
+  isAdded?: boolean
+  constructor(values?: Partial<DirectMessages>) {
+    Object.assign(this, values)
+    this[immerable] = true
+  }
+}
+
+export type DirectMessagesStore = DirectMessages
+
+export const initialState: DirectMessagesStore = {
+  users: {
+    '02dc8264c555d46b3f6b16f1e751e979ebc69e6df6a02e7d4074a5df981e507da2': {
+      nickname: 'holmes',
+      publicKey: '02dc8264c555d46b3f6b16f1e751e979ebc69e6df6a02e7d4074a5df981e507da2',
+      halfKey: '279e40e4ad5bc84f6cfcdb90465317e61255ba7ee78600179ea129a77e1bcef4'
+    }
+  },
+  conversations: {},
+  conversationsList: {},
+  privateKey: '',
+  publicKey: ''
+}
+
+const fetchUsers = createAction<{ usersList: { [key: string]: IUser } }>(actionTypes.FETCH_USERS)
+const setPublicKey = createAction<string>(actionTypes.SET_PUBLIC_KEY)
+const setPrivateKey = createAction<string>(actionTypes.SET_PRIVATE_KEY)
+const addConversation = createAction<IConversation>(actionTypes.ADD_CONVERSATION)
+const fetchConversations = createAction<{ conversationsList: { [key: string]: string } }>(
+  actionTypes.FETCH_CONVERSATIONS
+)
+
+export const actions = {
+  fetchUsers,
+  setPublicKey,
+  setPrivateKey,
+  addConversation,
+  fetchConversations
+}
+
+export type DirectMessagesActions = ActionsType<typeof actions>
+
+const generateDiffieHellman = () => async dispatch => {
+  const prime = 'b25dbea8c5f6c0feff269f88924a932639f8d8f937d19fa5874188258a63a373'
+  const generator = '02'
+
+  const dh = crypto.createDiffieHellman(prime, 'hex', generator, 'hex')
+  dh.generateKeys()
+  const privateKey = dh.getPrivateKey('hex')
+  const publicKey = dh.getPublicKey('hex')
+
+  await dispatch(actions.setPrivateKey(privateKey))
+  await dispatch(actions.setPublicKey(publicKey))
+}
+
+export const getPrivateConversations = () => dispatch => {
+  dispatch(directMessagesActions.getPrivateConversations())
+}
+
+const encodeMessage = (sharedSecret, message) => {
```

Comment:
Isn't it the same function as in socket.saga.ts?
Maybe it's a good idea to store crypto-functions (decodeMessage, encodeMessage, checkConversation) in separate file?

Reply:
Remember about types.